File: bindings/java/src/test/java/ai/arcships/aimux/ToolCallRepairTest.java
package ai.arcships.aimux;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import org.json.JSONArray;
import org.json.JSONObject;
import org.junit.jupiter.api.Test;

import java.io.File;
import java.io.IOException;
import java.time.Duration;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;
import java.util.stream.Collectors;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.junit.jupiter.api.Assertions.assertTimeoutPreemptively;

/**
 * Host-side tool-call repair (RFC-0035).
 *
 * <p>Two layers: a replay of the shared
 * {@code contract-tests/fixtures/tool-call-repair.json} over the three native
 * functions, plus public typed non-streaming/streaming integration and the
 * same-model nested-generation deadlock regression. Tests use only mock
 * providers, but need the native library on {@code java.library.path}.
 */
class ToolCallRepairTest {

    private static final String[] FIXTURE_CANDIDATES = {
        "../../contract-tests/fixtures/tool-call-repair.json", // gradle test cwd = bindings/java
        "contract-tests/fixtures/tool-call-repair.json",       // run from the repo root
        "../contract-tests/fixtures/tool-call-repair.json",
    };

    // ── shared fixture replay ────────────────────────────────────────────────

    @Test
    void contractFixtureReplays() throws Exception {
        JsonNode cases = loadFixture().get("cases");
        List<String> replayed = new ArrayList<>();
        for (JsonNode c : cases) {
            String name = c.get("name").asText();
            String expectedError = c.path("expected_error").asText(null);
            if (expectedError != null) {
                // Every error case in the fixture is InvalidArgument; if a new
                // variant appears, this pins it rather than silently passing.
                assertThat(expectedError).describedAs(name).isEqualTo("InvalidArgument");
                assertThatThrownBy(() -> invoke(c))
                    .describedAs(name)
                    .isInstanceOf(AimuxException.InvalidArgumentError.class);
            } else {
                assertThat(json(invoke(c))).describedAs(name).isEqualTo(c.get("expected"));
            }
            replayed.add(name);
        }
        assertThat(replayed).hasSize(cases.size());
    }

    private static String invoke(JsonNode testCase) {
        JsonNode in = testCase.get("input");
        String function = testCase.get("function").asText();
        String opts = text(in.get("opts"));
        if ("tool_call_repair_context".equals(function)) {
            return ToolCallRepairs.context(in.get("tool_call").toString(), text(in.get("prompt")), opts);
        }
        if ("apply_tool_call_repair".equals(function)) {
            return ToolCallRepairs.apply(in.get("tool_call").toString(), opts, in.get("reply").toString());
        }
        if ("apply_tool_call_repair_to_result".equals(function)) {
            return ToolCallRepairs.applyToResult(in.get("result").toString(), opts,
                in.get("tool_call_id").asText(), in.get("reply").toString());
        }
        throw new AssertionError("unknown fixture function: " + function);
    }

    // ── end to end through TypedModel (mock provider) ────────────────────────

    @Test
    void generateTextRepairsTheInvalidCallItGotBack() {
        try (MockProviderServer server = new MockProviderServer()) {
            server.setResponseBody(toolCallResponse("{\"city\":\"Tokyo\"}"));
            Types.GenerateTextOptions options = Types.GenerateTextOptions.builder()
                .tools(Collections.singletonList(strictWeatherTool()))
                .repairToolCall(context -> {
                    assertThat(context.getToolCall().getInput()).isEqualTo("{\"city\":\"Tokyo\"}");
                    return Types.RawToolCall.builder()
                        .toolCallId(context.getToolCall().getToolCallId())
                        .toolName(context.getToolCall().getToolName())
                        .input(context.getToolCall().getInput().replace("city", "location"))
                        .build();
                })
                .build();

            try (TypedModel model =
                     TypedModel.openaiWithBase("sk-test-fake-key", "gpt-4o", server.baseUrl())) {
                Types.GenerateTextResult result =
                    model.generateText("What is the weather in Tokyo?", options);

                Types.ToolCall call = result.getToolCalls().get(0);
                assertThat(call.getInvalid()).isNull();
                assertThat(call.getInput().get("location").asText()).isEqualTo("Tokyo");
                Types.ContentPart.ToolCall transcriptCall = result.getResponseMessages().stream()
                    .flatMap(message -> message.getContentParts().stream())
                    .filter(part -> part instanceof Types.ContentPart.ToolCall)
                    .map(part -> (Types.ContentPart.ToolCall) part)
                    .findFirst().orElseThrow(AssertionError::new);
                assertThat(transcriptCall.getInput().get("location").asText()).isEqualTo("Tokyo");
            }
        }
    }

    @Test
    void generateTextAsOpenAICarriesTheRepairedArguments() {
        try (MockProviderServer server = new MockProviderServer()) {
            server.setResponseBody(toolCallResponse("{\"city\":\"Tokyo\"}"));
            Types.GenerateTextOptions options = Types.GenerateTextOptions.builder()
                .tools(Collections.singletonList(strictWeatherTool()))
                .repairToolCall(context -> Types.RawToolCall.builder()
                    .toolCallId(context.getToolCall().getToolCallId())
                    .toolName(context.getToolCall().getToolName())
                    .input(context.getToolCall().getInput().replace("city", "location"))
                    .build())
                .build();

            try (TypedModel model =
                     TypedModel.openaiWithBase("sk-test-fake-key", "gpt-4o", server.baseUrl())) {
                // A ChatCompletion has no invalid marker: the native result is
                // repaired, then converted.
                Types.ChatCompletion completion =
                    model.generateTextAsOpenAI("What is the weather in Tokyo?", options);

                Types.ChatCompletionToolCall call =
                    completion.getChoices().get(0).getMessage().getToolCalls().get(0);
                assertThat(call.getFunction().getArguments()).isEqualTo("{\"location\":\"Tokyo\"}");
            }
        }
    }

    @Test
    void aStreamRepairFailingAtTheBoundaryStillDeliversTheCall() {
        try (MockProviderServer server = new MockProviderServer()) {
            server.setContentType("text/event-stream");
            server.setResponseBody(toolCallSse("{\"city\":\"Tokyo\"}"));
            // A replacement without a tool_call_id is not a RawToolCall on the
            // wire: the library rejects the reply itself (not the hook failing).
            Types.GenerateTextOptions options = Types.GenerateTextOptions.builder()
                .tools(Collections.singletonList(strictWeatherTool()))
                .repairToolCall(context -> Types.RawToolCall.builder()
                    .toolCallId(null)
                    .toolName(context.getToolCall().getToolName())
                    .input("{\"location\":\"Tokyo\"}")
                    .build())
                .build();

            try (TypedModel model =
                     TypedModel.openaiWithBase("sk-test-fake-key", "gpt-4o", server.baseUrl())) {
                List<Types.StreamPart> parts = new ArrayList<>();
                List<String> errors = new ArrayList<>();
                model.streamText("What is the weather in Tokyo?", options,
                    parts::add, () -> { }, errors::add);

                assertThat(errors).hasSize(1);
                assertThat(errors.get(0)).startsWith("failed to repair tool call:");
                List<Types.StreamPart.ToolCall> calls = parts.stream()
                    .filter(p -> p instanceof Types.StreamPart.ToolCall)
                    .map(p -> (Types.StreamPart.ToolCall) p)
                    .collect(Collectors.toList());
                assertThat(calls).hasSize(1);
                assertThat(calls.get(0).getInvalid()).isTrue();
            }
        }
    }

    @Test
    void streamTextStreamDeliversTheRepairedToolCallPart() {
        try (MockProviderServer server = new MockProviderServer()) {
            server.setContentType("text/event-stream");
            server.setResponseBody(toolCallSse("{\"city\":\"Tokyo\"}"));
            Types.GenerateTextOptions options = Types.GenerateTextOptions.builder()
                .tools(Collections.singletonList(strictWeatherTool()))
                .repairToolCall(context -> Types.RawToolCall.builder()
                    .toolCallId(context.getToolCall().getToolCallId())
                    .toolName(context.getToolCall().getToolName())
                    .input(context.getToolCall().getInput().replace("city", "location"))
                    .build())
                .build();

            try (TypedModel model =
                     TypedModel.openaiWithBase("sk-test-fake-key", "gpt-4o", server.baseUrl())) {
                List<Types.StreamPart> parts =
                    model.streamTextStream("What is the weather in Tokyo?", options)
                        .collect(Collectors.toList());

                List<Types.StreamPart.ToolCall> calls = parts.stream()
                    .filter(p -> p instanceof Types.StreamPart.ToolCall)
                    .map(p -> (Types.StreamPart.ToolCall) p)
                    .collect(Collectors.toList());
                assertThat(calls).hasSize(1);
                assertThat(calls.get(0).getInvalid()).isNull();
                assertThat(calls.get(0).getInput().get("location").asText()).isEqualTo("Tokyo");

                // Argument deltas are forwarded verbatim, repair or not.
                assertThat(parts.stream()
                    .filter(p -> p instanceof Types.StreamPart.ToolInputDelta)
                    .map(p -> ((Types.StreamPart.ToolInputDelta) p).getDelta())
                    .collect(Collectors.joining())).contains("city");
            }
        }
    }

    @Test
    void theRepairLoopHandlesThrowNullValidCallAndNoToolSet() {
        try (MockProviderServer server = new MockProviderServer();
             TypedModel model =
                 TypedModel.openaiWithBase("sk-test-fake-key", "gpt-4o", server.baseUrl())) {
            final AtomicInteger hookCalls = new AtomicInteger();
            ToolCallRepair countingNull = context -> {
                hookCalls.incrementAndGet();
                return null;
            };

            // The hook throws: ToolCallRepair { original_error, cause: Other }.
            server.setResponseBody(toolCallResponse("{\"city\":\"Tokyo\"}"));
            Types.ToolCall call = model.generateText("What is the weather in Tokyo?",
                Types.GenerateTextOptions.builder()
                    .tools(Collections.singletonList(strictWeatherTool()))
                    .repairToolCall(context -> {
                        hookCalls.incrementAndGet();
                        throw new IllegalStateException("repair model unavailable");
                    })
                    .build()).getToolCalls().get(0);
            assertThat(call.getInvalid()).isTrue();
            JsonNode failure = call.getError().path("ToolCallRepair");
            assertThat(failure.path("cause").path("Other").asText()).isEqualTo("repair model unavailable");
            assertThat(failure.path("original_error").has("InvalidToolInput")).isTrue();
            assertThat(hookCalls.get()).isEqualTo(1);

            // The hook returns null: the ORIGINAL error, not wrapped in ToolCallRepair.
            server.setResponseBody(toolCallResponse("{\"city\":\"Tokyo\"}"));
            call = model.generateText("What is the weather in Tokyo?",
                Types.GenerateTextOptions.builder()
                    .tools(Collections.singletonList(strictWeatherTool()))
                    .repairToolCall(countingNull)
                    .build()).getToolCalls().get(0);
            assertThat(call.getInvalid()).isTrue();
            assertThat(call.getError().has("ToolCallRepair")).isFalse();
            assertThat(call.getError().has("InvalidToolInput")).isTrue();
            assertThat(hookCalls.get()).isEqualTo(2);

            // A valid call never reaches the hook.
            server.setResponseBody(toolCallResponse("{\"location\":\"Tokyo\"}"));
            call = model.generateText("What is the weather in Tokyo?",
                Types.GenerateTextOptions.builder()
                    .tools(Collections.singletonList(strictWeatherTool()))
                    .repairToolCall(countingNull)
                    .build()).getToolCalls().get(0);
            assertThat(call.getInvalid()).isNull();
            assertThat(hookCalls.get()).isEqualTo(2);

            // No tool set: the native context answers null, so the host skips the hook.
            server.setResponseBody(toolCallResponse("{\"city\":\"Tokyo\"}"));
            call = model.generateText("What is the weather in Tokyo?",
                Types.GenerateTextOptions.builder().repairToolCall(countingNull).build())
                .getToolCalls().get(0);
            assertThat(call.getInvalid()).isTrue();
            assertThat(hookCalls.get()).isEqualTo(2);
        }
    }

    @Test
    void aStreamHookMayGenerateWithTheSameModelWhileCloseIsPending() {
        // Regression: the stream holds Model's (fair) read lock for its whole
        // duration, a queued close() waits for it, and a hook generating with
        // the SAME model used to queue behind that writer — S waits for R waits
        // for C waits for S. The stream thread lends its read hold to the repair
        // worker so the hook's call takes no lock at all.
        assertTimeoutPreemptively(Duration.ofSeconds(30), () -> {
            try (MockProviderServer server = new MockProviderServer()) {
                server.setContentType("text/event-stream");
                server.setResponses(toolCallSse("{\"city\":\"Tokyo\"}"),
                                    textResponse("{\"location\":\"Tokyo\"}"));

                final Model raw =
                    Model.openaiWithBase("sk-test-fake-key", "gpt-4o", server.baseUrl());
                final TypedModel model = TypedModel.of(raw);
                final CountDownLatch hookStarted = new CountDownLatch(1);
                final AtomicReference<Throwable> closerFailure = new AtomicReference<>();
                Thread closer = new Thread(() -> {
                    try {
                        hookStarted.await();
                        raw.close();
                    } catch (Throwable t) {
                        closerFailure.set(t);
                    }
                }, "closer");
                closer.start();

                Types.GenerateTextOptions options = Types.GenerateTextOptions.builder()
                    .tools(Collections.singletonList(strictWeatherTool()))
                    .repairToolCall(context -> {
                        hookStarted.countDown();
                        // Let the closer actually enqueue for the write lock;
                        // without it the race the test reproduces may not happen.
                        Thread.sleep(200);
                        // The mock server's content type is global, so flip it
                        // now that the SSE response has already been written.
                        server.setContentType("application/json");
                        return Types.RawToolCall.builder()
                            .toolCallId(context.getToolCall().getToolCallId())
                            .toolName(context.getToolCall().getToolName())
                            .input(model.generateText(
                                "fix these arguments: " + context.getToolCall().getInput())
                                .getText())
                            .build();
                    })
                    .build();

                List<Types.StreamPart> parts = new ArrayList<>();
                model.streamText("What is the weather in Tokyo?", options,
                    parts::add, () -> {}, error -> { throw new AssertionError(error); });

                closer.join(10_000);
                assertThat(closer.isAlive()).isFalse();
                assertThat(closerFailure.get()).isNull();

                List<Types.StreamPart.ToolCall> calls = parts.stream()
                    .filter(p -> p instanceof Types.StreamPart.ToolCall)
                    .map(p -> (Types.StreamPart.ToolCall) p)
                    .collect(Collectors.toList());
                assertThat(calls).hasSize(1);
                assertThat(calls.get(0).getInvalid()).isNull();
                assertThat(calls.get(0).getInput().get("location").asText()).isEqualTo("Tokyo");
                raw.close(); // idempotent; the closer already did it
            }
        });
    }

    /** A plain OpenAI text response whose content is {@code text}. */
    private static String textResponse(String text) {
        return new JSONObject()
            .put("id", "chatcmpl-repair")
            .put("model", "gpt-4o")
            .put("choices", new JSONArray().put(new JSONObject()
                .put("message", new JSONObject().put("role", "assistant").put("content", text))
                .put("finish_reason", "stop")))
            .put("usage", new JSONObject()
                .put("prompt_tokens", 4).put("completion_tokens", 4).put("total_tokens", 8))
            .toString();
    }

    /** {@code get_weather}, schema-strict so a {@code city} argument is rejected. */
    private static Types.Tool strictWeatherTool() {
        ObjectNode schema = Types.AimuxJson.MAPPER.createObjectNode();
        schema.put("type", "object");
        schema.set("properties", Types.AimuxJson.MAPPER.createObjectNode()
            .set("location", Types.AimuxJson.MAPPER.createObjectNode().put("type", "string")));
        schema.set("required", Types.AimuxJson.MAPPER.createArrayNode().add("location"));
        schema.put("additionalProperties", false);
        return Types.Tool.Function.builder().name("get_weather").inputSchema(schema).build();
    }

    private static String toolCallResponse(String arguments) {
        return new JSONObject()
            .put("id", "chatcmpl-tc")
            .put("model", "gpt-4o")
            .put("choices", new JSONArray().put(new JSONObject()
                .put("message", new JSONObject()
                    .put("role", "assistant")
                    .put("content", JSONObject.NULL)
                    .put("tool_calls", new JSONArray().put(new JSONObject()
                        .put("id", "call_abc")
                        .put("type", "function")
                        .put("function", new JSONObject()
                            .put("name", "get_weather")
                            .put("arguments", arguments)))))
                .put("finish_reason", "tool_calls")))
            .put("usage", new JSONObject()
                .put("prompt_tokens", 20).put("completion_tokens", 10).put("total_tokens", 30))
            .toString();
    }

    private static String toolCallSse(String arguments) {
        StringBuilder sb = new StringBuilder();
        sb.append("data: ").append(new JSONObject()
            .put("id", "1").put("model", "gpt-4o")
            .put("choices", new JSONArray().put(new JSONObject().put("delta", new JSONObject()
                .put("role", "assistant")
                .put("tool_calls", new JSONArray().put(new JSONObject()
                    .put("index", 0).put("id", "call_xyz").put("type", "function")
                    .put("function", new JSONObject().put("name", "get_weather")
                        .put("arguments", "")))))))).append("\n\n");
        sb.append("data: ").append(new JSONObject()
            .put("id", "1").put("model", "gpt-4o")
            .put("choices", new JSONArray().put(new JSONObject().put("delta", new JSONObject()
                .put("tool_calls", new JSONArray().put(new JSONObject()
                    .put("index", 0)
                    .put("function", new JSONObject().put("arguments", arguments)))))))).append("\n\n");
        sb.append("data: ").append(new JSONObject()
            .put("id", "1").put("model", "gpt-4o")
            .put("choices", new JSONArray().put(new JSONObject()
                .put("delta", new JSONObject()).put("finish_reason", "tool_calls")))
            .put("usage", new JSONObject()
                .put("prompt_tokens", 5).put("completion_tokens", 2).put("total_tokens", 7)))
            .append("\n\n");
        sb.append("data: [DONE]\n\n");
        return sb.toString();
    }

    private static JsonNode json(String raw) throws IOException {
        return Types.AimuxJson.MAPPER.readTree(raw);
    }

    private static String text(JsonNode node) {
        return node == null || node.isNull() ? null : node.toString();
    }

    private static JsonNode loadFixture() throws IOException {
        for (String candidate : FIXTURE_CANDIDATES) {
            File f = new File(candidate);
            if (f.isFile()) {
                return Types.AimuxJson.MAPPER.readTree(f);
            }
        }
        throw new IllegalStateException(
            "cannot find tool-call-repair.json; tried " + Arrays.toString(FIXTURE_CANDIDATES)
                + " from " + new File(".").getAbsolutePath());
    }
}

File: bindings/kotlin/src/test/kotlin/ai/arcships/aimux/ToolCallRepairTest.kt
package ai.arcships.aimux

import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.assertTimeoutPreemptively
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.io.File
import java.time.Duration
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

// ─────────────────────────────────────────────────────────────────────────────
// Host-side tool-call repair (RFC-0035).
//
// Two halves:
//   1. the shared `tool-call-repair.json` fixture replayed through the three
//      native entry points (pure functions — no model, no server),
//   2. end-to-end through TypedModel against MockProviderServer: the provider
//      returns arguments that violate the tool schema, Core marks the call
//      invalid, and the repair function runs on the JVM.
// ─────────────────────────────────────────────────────────────────────────────

class ToolCallRepairTest {

    private lateinit var server: MockProviderServer

    @BeforeEach
    fun setUp() {
        server = MockProviderServer()
    }

    @AfterEach
    fun tearDown() {
        server.stop()
    }

    // ── shared fixture ──────────────────────────────────────────────────

    private val fixtureCandidates = listOf(
        "../../contract-tests/fixtures/tool-call-repair.json", // gradle test cwd = bindings/kotlin
        "contract-tests/fixtures/tool-call-repair.json",
        "../contract-tests/fixtures/tool-call-repair.json",
    )

    private fun fixtureCases(): List<JsonObject> {
        val file = fixtureCandidates.map(::File).firstOrNull { it.isFile }
            ?: throw IllegalStateException(
                "cannot find tool-call-repair.json; tried $fixtureCandidates from ${File(".").absolutePath}"
            )
        val cases = AimuxJson.parseToJsonElement(file.readText()).jsonObject["cases"]!!
        return cases.let { it as kotlinx.serialization.json.JsonArray }
            .map { it.jsonObject }
    }

    /** `opts` may be JSON null in the fixture, which is the binding's "defaults". */
    private fun optsOf(input: JsonObject): String? =
        input["opts"]?.takeIf { it !is JsonNull }?.toString()

    @Test
    fun `all shared contract fixture cases replay`() {
        fixtureCases().forEach { case ->
            val name = case["name"]!!.jsonPrimitive.content
            val input = case["input"]!!.jsonObject
            val invoke = {
                when (val function = case["function"]!!.jsonPrimitive.content) {
                    "tool_call_repair_context" -> toolCallRepairContext(
                        input["tool_call"]!!.toString(), input["prompt"]!!.toString(), optsOf(input))
                    "apply_tool_call_repair" -> applyToolCallRepair(
                        input["tool_call"]!!.toString(), optsOf(input), input["reply"]!!.toString())
                    "apply_tool_call_repair_to_result" -> applyToolCallRepairToResult(
                        input["result"]!!.toString(), optsOf(input),
                        input["tool_call_id"]!!.jsonPrimitive.content, input["reply"]!!.toString())
                    else -> throw AssertionError("unknown fixture function: $function")
                }
            }
            if (case["expected_error"] != null) {
                assertThat(case["expected_error"]!!.jsonPrimitive.content).isEqualTo("InvalidArgument")
                assertThatThrownBy(invoke).describedAs(name).isInstanceOf(InvalidArgumentError::class.java)
            } else {
                assertThat(AimuxJson.parseToJsonElement(invoke())).describedAs(name)
                    .isEqualTo(case["expected"]!!)
            }
        }
    }

    // ── end to end ──────────────────────────────────────────────────────

    /** `city` is required; the canned response sends `town` instead. */
    private val weatherTool: Tool = Tool.Function(
        name = "weather",
        inputSchema = AimuxJson.parseToJsonElement(
            """{"type":"object","properties":{"city":{"type":"string"}},"required":["city"],"additionalProperties":false}"""
        ),
    )

    private fun openAiToolCall(arguments: String, name: String = "weather"): String =
        """
        {"id":"chatcmpl-1","model":"gpt-4o","choices":[{"message":{"role":"assistant","content":null,
         "tool_calls":[{"id":"call-1","type":"function","function":{"name":"$name","arguments":${
            JsonPrimitive(arguments)
        }}}]},"finish_reason":"tool_calls"}],
         "usage":{"prompt_tokens":1,"completion_tokens":1,"total_tokens":2}}
        """.trimIndent()

    private fun model(): TypedModel = TypedModel.openai("sk-test-fake-key", "gpt-4o", server.baseUrl)

    private fun options(repair: RepairToolCall?): GenerateTextOptions =
        GenerateTextOptions(tools = listOf(weatherTool), repairToolCall = repair)

    @Test
    fun `a repaired call becomes valid in both tool_calls and response_messages`() {
        server.responseBody = openAiToolCall("""{"town":"Singapore"}""")

        model().use { model ->
            val result = model.generateText(
                "weather in Singapore?",
                options { context ->
                    // The repair function sees the RAW argument text, not a parsed object.
                    assertThat(context.toolCall.input).isEqualTo("""{"town":"Singapore"}""")
                    assertThat(context.inputSchema.jsonObject["required"].toString()).contains("city")
                    assertThat(context.messages).hasSize(1)
                    context.toolCall.copy(input = """{"city":"Singapore"}""")
                },
            )

            val call = result.toolCalls.single()
            assertThat(call.invalid).isNotEqualTo(true)
            assertThat(call.error).isNull()
            assertThat(call.input.jsonObject["city"]?.jsonPrimitive?.content).isEqualTo("Singapore")

            // The transcript must agree, or the next turn replays bad arguments.
            val part = result.responseMessages
                .flatMap { it.contentParts ?: emptyList() }
                .filterIsInstance<ContentPart.ToolCall>()
                .single()
            assertThat(part.toolCallId).isEqualTo("call-1")
            assertThat(part.input.jsonObject["city"]?.jsonPrimitive?.content).isEqualTo("Singapore")
        }
    }

    /** The nested generation a repair hook performs: valid arguments as text. */
    private val repairCompletion: String =
        """{"id":"chatcmpl-2","model":"gpt-4o","choices":[{"message":{"role":"assistant",
           "content":"{\"city\":\"Singapore\"}"},"finish_reason":"stop"}],
           "usage":{"prompt_tokens":1,"completion_tokens":1,"total_tokens":2}}"""

    /** SSE: tool-call start, one arguments delta violating the schema, finish. */
    private val weatherSse: String = buildString {
        append("""data: {"id":"1","model":"gpt-4o","choices":[{"delta":{"role":"assistant","tool_calls":""")
        append("""[{"index":0,"id":"call-1","type":"function","function":{"name":"weather","arguments":""}}]}}]}""")
        append("\n\n")
        append("""data: {"id":"1","model":"gpt-4o","choices":[{"delta":{"tool_calls":""")
        append("""[{"index":0,"function":{"arguments":"{\"town\":\"Singapore\"}"}}]}}]}""")
        append("\n\n")
        append("""data: {"id":"1","model":"gpt-4o","choices":[{"delta":{},"finish_reason":"tool_calls"}]}""")
        append("\n\ndata: [DONE]\n\n")
    }

    @Test
    fun `streaming replaces the tool-call part and leaves the deltas alone`() {
        server.contentType = "text/event-stream"
        server.responseBody = weatherSse

        model().use { model ->
            val parts = model.streamTextSequence(
                "weather in Singapore?",
                options { it.toolCall.copy(input = """{"city":"Singapore"}""") },
            ).toList()

            val call = parts.filterIsInstance<StreamPart.ToolCall>().single()
            assertThat(call.invalid).isNotEqualTo(true)
            assertThat(call.input.jsonObject["city"]?.jsonPrimitive?.content).isEqualTo("Singapore")

            // Deltas are the provider's text, forwarded untouched.
            val delta = parts.filterIsInstance<StreamPart.ToolInputDelta>().single()
            assertThat(delta.delta).isEqualTo("""{"town":"Singapore"}""")
        }
    }

    @Test
    fun `a streaming hook may call the SAME model while another thread closes it`() {
        // The lock is fair, so a reader that arrives after close() has queued
        // for the write lock waits for it. Without the lent read hold this is a
        // three-way deadlock: stream thread (holds read, waits for the hook) ←
        // closer (waits for the write lock) ← hook (waits for a read lock).
        server.contentType = "text/event-stream"
        server.setResponses(weatherSse, repairCompletion)

        val model = model()
        val hookStarted = CountDownLatch(1)
        val closer = Thread {
            hookStarted.await()
            model.close()
        }
        closer.isDaemon = true
        // assertTimeoutPreemptively, not @Timeout: a regression deadlocks, and
        // the default @Timeout mode only checks after the body returns.
        assertTimeoutPreemptively(Duration.ofSeconds(30)) {
            closer.start()
            val parts = model.streamTextSequence(
                "weather in Singapore?",
                options { context ->
                    // The stream response is already fully written; the nested
                    // completion is plain JSON off the same one-server queue.
                    server.contentType = "application/json"
                    hookStarted.countDown()
                    // Let close() reach the write lock and queue ahead of this
                    // thread's read acquisition — that ordering is the bug, and
                    // without the sleep the hook wins the race and proves
                    // nothing.
                    Thread.sleep(250)
                    val fixed = model.generateText("Fix ${context.toolCall.input}")
                    context.toolCall.copy(input = fixed.text)
                },
            ).toList()

            val call = parts.filterIsInstance<StreamPart.ToolCall>().single()
            assertThat(call.error).isNull()
            assertThat(call.invalid).isNotEqualTo(true)
            assertThat(call.input.jsonObject["city"]?.jsonPrimitive?.content).isEqualTo("Singapore")

            closer.join(TimeUnit.SECONDS.toMillis(10))
            assertThat(closer.isAlive).describedAs("close() never completed").isFalse()
        }
    }

}
